### PR TITLE
add rectangle stencil

### PR DESCRIPTION
**Pull request by rafaqz**
@alex-s-gardner this adds a `Rectangle` stencil.

Turns out its not so easy to switch from `Window` to `Rectangle` in `Kernel`!

What do we do if the array is already offset around zero? e.g an OffsetArray

### CODE REVIEW

**Comment by alex-s-gardner** [comment]
When trying to specify a 3D rectangle I get this error and I'm not sure why All offsets must be the length `N` as I'm just defining the distance from center for a single coordinate. 

```julia
julia> Stencils.Rectangle((1,1),(1,1),(6,5))
ERROR: AssertionError: All offsets must be the length `N` of 3, got ((1, 1), (1, 1), (6, 5))
Stacktrace:
 [1] Stencils.Rectangle{((1, 1), (1, 1), (6, 5)), 6, 3, 0, Union{}}(neighbors::SVector{0, Union{}})
   @ Stencils ~/.julia/packages/Stencils/WyCBg/src/stencils/rectangle.jl:32
 [2] (Stencils.Rectangle{((1, 1), (1, 1), (6, 5)), 6, 3, 0})(neighbors::SVector{0, Union{}})
   @ Stencils ~/.julia/packages/Stencils/WyCBg/src/stencils/rectangle.jl:36
 [3] (Stencils.Rectangle{((1, 1), (1, 1), (6, 5)), 6, 3, 0})()
   @ Stencils ~/.julia/packages/Stencils/WyCBg/src/stencils/rectangle.jl:38
 [4] (Stencils.Rectangle{((1, 1), (1, 1), (6, 5))})()
   @ Stencils ~/.julia/packages/Stencils/WyCBg/src/stencils/rectangle.jl:46
 [5] Stencils.Rectangle(offsets::Tuple{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{Int64, Int64}})
   @ Stencils ~/.julia/packages/Stencils/WyCBg/src/stencils/rectangle.jl:49
 [6] Stencils.Rectangle(::Tuple{Int64, Int64}, ::Tuple{Int64, Int64}, ::Vararg{Tuple{Int64, Int64}})
   @ Stencils ~/.julia/packages/Stencils/WyCBg/src/stencils/rectangle.jl:48
 [7] top-level scope
   @ REPL[9]:1
```

**Comment by rafaqz** [comment]
Ah sorry I copied the code from Positional I must have forgotten to delete that check

File: src/stencils/rectangle.jl [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+Positional{((0, -1), (2, 1), (-1, 1), (0, 1)), 2, 2, 4, Nothing}
+   ▄ 
+ ▀ ▀ 
+   ▀ 
+```
+"""
+struct Rectangle{O,R,N,L,T} <: Stencil{R,N,L,T}
+    neighbors::SVector{L,T}
+    function Rectangle{O,R,N,L,T}(neighbors::SVector{L,T}) where {O,R,N,L,T} 
+        @assert all(map(o -> length(o) == N, O)) "All offsets must be the length `N` of $N, got $O"
```
Comment (alex-s-gardner) [inline src/stencils/rectangle.jl:32]: remove assertion

Comment (rafaqz) [reply src/stencils/rectangle.jl:32]: or change it to checking the length == 2 ...

Comment (rafaqz) [reply src/stencils/rectangle.jl:32]: this is pushed

**Comment by alex-s-gardner** [comment]
Rectangle should probably be exported

**Comment by alex-s-gardner** [comment]
Rectangle does not work with GPU

Test window stencil (works)
```julia 

using Stencils;
using Metal;

# test window stencil
r = rand(Float32, 2^6, 2^6);
k = rand(Float32.(1:10), (3, 3));

s = Window(1);
sk = Stencils.Kernel(s,k)

A = StencilArray(r, sk);

println("kernel - rectangle")
begin
    B = similar(A)
    println("mapstencil! - CPU")
    display(@benchmark mapstencil!(kernelproduct, $B, $A))
    println("mapstencil! - GPU")
    sk2 = Stencils.Kernel(s, MtlArray(k))
    C = StencilArray(MtlArray(r), sk2)
    D = similar(C)
    display(@benchmark mapstencil!(kernelproduct, $D, $C))
end
```

Test GPU [does not work]
```
# test rectangular stencil
s = Stencils.Rectangle((-1,1), (-1,1))
sk = Stencils.Kernel(s, k)
A = StencilArray(r, sk);

println("kernel - rectangle")
begin
    B = similar(A)
    println("mapstencil! - CPU")
    display(@benchmark mapstencil!(kernelproduct, $B, $A))
    println("mapstencil! - GPU")
    sk2 = Stencils.Kernel(s, MtlArray(k))
    C = StencilArray(MtlArray(r), sk2)
    D = similar(C)
    display(@benchmark mapstencil!(kernelproduct, $D, $C))
end
```
returns the following error
```julia
mapstencil! - GPU
ERROR: InvalidIRError: compiling MethodInstance for Stencils.gpu_mapstencil_kerneln0!(::KernelAbstractions.CompilerMetadata{…}, ::typeof(kernelproduct), ::MtlDeviceMatrix{…}, ::StencilArray{…}) resulted in invalid LLVM IR
Reason: unsupported call to an unknown function (call to julia.new_gc_frame)
Stacktrace:
 [1] multiple call sites
   @ unknown:0
Reason: unsupported call to an unknown function (call to julia.push_gc_frame)
Stacktrace:
 [1] multiple call sites
   @ unknown:0
Reason: unsupported call through a literal pointer (call to ijl_alloc_array_2d)
Stacktrace:
  [1] Array
    @ ./boot.jl:477
  [2] Array
    @ ./boot.jl:485
  [3] similar
    @ ./abstractarray.jl:839
  [4] similar
    @ ./abstractarray.jl:836
  [5] _similar_for
    @ ./array.jl:715
  [6] _collect
    @ ./array.jl:857
  [7] collect_similar
    @ ./array.jl:761
  [8] map
    @ ./abstractarray.jl:3274
  [9] offsets
    @ ~/.julia/packages/Stencils/AwhBN/src/stencils/rectangle.jl:35
 [10] multiple call sites
    @ unknown:0
Reason: unsupported call to an unknown function (call to julia.get_gc_frame_slot)
Stacktrace:
 [1] multiple call sites
   @ unknown:0
Reason: unsupported dynamic function invocation (call to dimension_mismatch_fail(SA::Type, a::AbstractArray) @ StaticArrays ~/.julia/packages/StaticArrays/PLKkM/src/convert.jl:195)
Stacktrace:
 [1] convert
   @ ~/.julia/packages/StaticArrays/PLKkM/src/convert.jl:201
 [2] StaticArray
   @ ~/.julia/packages/StaticArrays/PLKkM/src/convert.jl:174
 [3] offsets
   @ ~/.julia/packages/Stencils/AwhBN/src/stencils/rectangle.jl:35
 [4] multiple call sites
   @ unknown:0
Reason: unsupported call to an unknown function (call to julia.pop_gc_frame)
Stacktrace:
 [1] multiple call sites
   @ unknown:0
Hint: catch this exception as `err` and call `code_typed(err; interactive = true)` to introspect the erronous code with Cthulhu.jl
Stacktrace:
  [1] check_ir(job::GPUCompiler.CompilerJob{GPUCompiler.MetalCompilerTarget, Metal.MetalCompilerParams}, args::LLVM.Module)
    @ GPUCompiler ~/.julia/packages/GPUCompiler/2mJjc/src/validation.jl:147
  [2] macro expansion
    @ ~/.julia/packages/GPUCompiler/2mJjc/src/driver.jl:440 [inlined]
  [3] macro expansion
    @ ~/.julia/packages/TimerOutputs/RsWnF/src/TimerOutput.jl:253 [inlined]
  [4] macro expansion
    @ ~/.julia/packages/GPUCompiler/2mJjc/src/driver.jl:439 [inlined]
  [5] 
    @ GPUCompiler ~/.julia/packages/GPUCompiler/2mJjc/src/utils.jl:92
  [6] emit_llvm
    @ ~/.julia/packages/GPUCompiler/2mJjc/src/utils.jl:86 [inlined]
  [7] 
    @ GPUCompiler ~/.julia/packages/GPUCompiler/2mJjc/src/driver.jl:129
  [8] codegen
    @ ~/.julia/packages/GPUCompiler/2mJjc/src/driver.jl:110 [inlined]
  [9] 
    @ GPUCompiler ~/.julia/packages/GPUCompiler/2mJjc/src/driver.jl:106
 [10] compile
    @ ~/.julia/packages/GPUCompiler/2mJjc/src/driver.jl:98 [inlined]
 [11] #45
    @ ~/.julia/packages/Metal/lnkVP/src/compiler/compilation.jl:57 [inlined]
 [12] JuliaContext(f::Metal.var"#45#46"{GPUCompiler.CompilerJob{GPUCompiler.MetalCompilerTarget, Metal.MetalCompilerParams}})
    @ GPUCompiler ~/.julia/packages/GPUCompiler/2mJjc/src/driver.jl:47
 [13] compile(job::GPUCompiler.CompilerJob)
    @ Metal ~/.julia/packages/Metal/lnkVP/src/compiler/compilation.jl:56
 [14] actual_compilation(cache::Dict{…}, src::Core.MethodInstance, world::UInt64, cfg::GPUCompiler.CompilerConfig{…}, compiler::typeof(Metal.compile), linker::typeof(Metal.link))
    @ GPUCompiler ~/.julia/packages/GPUCompiler/2mJjc/src/execution.jl:125
 [15] cached_compilation(cache::Dict{…}, src::Core.MethodInstance, cfg::GPUCompiler.CompilerConfig{…}, compiler::Function, linker::Function)
    @ GPUCompiler ~/.julia/packages/GPUCompiler/2mJjc/src/execution.jl:103
 [16] macro expansion
    @ ~/.julia/packages/Metal/lnkVP/src/compiler/execution.jl:162 [inlined]
 [17] macro expansion
    @ ./lock.jl:267 [inlined]
 [18] mtlfunction(f::typeof(Stencils.gpu_mapstencil_kerneln0!), tt::Type{Tuple{…}}; name::Nothing, kwargs::@Kwargs{})
    @ Metal ~/.julia/packages/Metal/lnkVP/src/compiler/execution.jl:157
 [19] mtlfunction(f::typeof(Stencils.gpu_mapstencil_kerneln0!), tt::Type{Tuple{…}})
    @ Metal ~/.julia/packages/Metal/lnkVP/src/compiler/execution.jl:155
 [20] macro expansion
    @ ~/.julia/packages/Metal/lnkVP/src/compiler/execution.jl:77 [inlined]
 [21] (::KernelAbstractions.Kernel{…})(::Function, ::Vararg{…}; ndrange::Nothing, workgroupsize::Nothing)
    @ Metal.MetalKernels ~/.julia/packages/Metal/lnkVP/src/MetalKernels.jl:105
 [22] Kernel
    @ ~/.julia/packages/Metal/lnkVP/src/MetalKernels.jl:101 [inlined]
 [23] mapstencil!(::Function, ::MtlMatrix{…}, ::StencilArray{…})
    @ Stencils ~/.julia/packages/Stencils/AwhBN/src/mapstencil.jl:81
 [24] var"##core#325"(D#319::MtlMatrix{…}, C#320::StencilArray{…})
    @ Main ~/.julia/packages/BenchmarkTools/QAv7m/src/execution.jl:547
 [25] var"##sample#326"(::Tuple{MtlMatrix{…}, StencilArray{…}}, __params::BenchmarkTools.Parameters)
    @ Main ~/.julia/packages/BenchmarkTools/QAv7m/src/execution.jl:556
 [26] _run(b::BenchmarkTools.Benchmark, p::BenchmarkTools.Parameters; verbose::Bool, pad::String, kwargs::@Kwargs{…})
    @ BenchmarkTools ~/.julia/packages/BenchmarkTools/QAv7m/src/execution.jl:109
 [27] #invokelatest#2
    @ ./essentials.jl:889 [inlined]
 [28] invokelatest
    @ ./essentials.jl:884 [inlined]
 [29] #run_result#45
    @ ~/.julia/packages/BenchmarkTools/QAv7m/src/execution.jl:41 [inlined]
 [30] run_result
    @ ~/.julia/packages/BenchmarkTools/QAv7m/src/execution.jl:40 [inlined]
 [31] run(b::BenchmarkTools.Benchmark, p::BenchmarkTools.Parameters; progressid::Nothing, nleaves::Float64, ndone::Float64, kwargs::@Kwargs{…})
    @ BenchmarkTools ~/.julia/packages/BenchmarkTools/QAv7m/src/execution.jl:134
 [32] run (repeats 2 times)
    @ BenchmarkTools ~/.julia/packages/BenchmarkTools/QAv7m/src/execution.jl:126 [inlined]
 [33] #warmup#54
    @ BenchmarkTools ~/.julia/packages/BenchmarkTools/QAv7m/src/execution.jl:189 [inlined]
 [34] warmup(item::BenchmarkTools.Benchmark)
    @ BenchmarkTools ~/.julia/packages/BenchmarkTools/QAv7m/src/execution.jl:188
 [35] macro expansion
    @ ~/.julia/packages/BenchmarkTools/QAv7m/src/execution.jl:432 [inlined]
 [36] top-level scope
    @ ~/Documents/GitHub/ItsLivePlayground.jl/src/Test_Stencils:104
Some type information was truncated. Use `show(err)` to see complete types.
```

**Comment by rafaqz** [comment]
Argg yeah everything has to be **perfect** for the gpu, no type issues, unions... there is some pain involved in getting those benchmarks...

**Comment by rafaqz** [comment]
This should work but I can't test on a GPU right now

**Comment by alex-s-gardner** [comment]
Looks like the error persists :-( :

```julia
ERROR: InvalidIRError: compiling MethodInstance for Stencils.gpu_mapstencil_kerneln0!(::KernelAbstractions.CompilerMetadata{…}, ::typeof(kernelproduct), ::MtlDeviceMatrix{…}, ::StencilArray{…}) resulted in invalid LLVM IR
Reason: unsupported call to an unknown function (call to julia.new_gc_frame)
Stacktrace:
 [1] multiple call sites
   @ unknown:0
Reason: unsupported call to an unknown function (call to julia.push_gc_frame)
Stacktrace:
 [1] multiple call sites
   @ unknown:0
Reason: unsupported call through a literal pointer (call to ijl_alloc_array_2d)
Stacktrace:
  [1] Array
    @ ./boot.jl:477
  [2] Array
    @ ./boot.jl:485
  [3] similar
    @ ./abstractarray.jl:839
  [4] similar
    @ ./abstractarray.jl:836
  [5] _similar_for
    @ ./array.jl:715
  [6] _collect
    @ ./array.jl:857
  [7] collect_similar
    @ ./array.jl:761
  [8] map
    @ ./abstractarray.jl:3274
  [9] offsets
    @ ~/.julia/packages/Stencils/FYZFU/src/stencils/rectangle.jl:35
 [10] multiple call sites
    @ unknown:0
Reason: unsupported call to an unknown function (call to julia.get_gc_frame_slot)
Stacktrace:
 [1] multiple call sites
   @ unknown:0
Reason: unsupported dynamic function invocation (call to dimension_mismatch_fail(SA::Type, a::AbstractArray) @ StaticArrays ~/.julia/packages/StaticArrays/PLKkM/src/convert.jl:195)
Stacktrace:
 [1] convert
   @ ~/.julia/packages/StaticArrays/PLKkM/src/convert.jl:201
 [2] StaticArray
   @ ~/.julia/packages/StaticArrays/PLKkM/src/convert.jl:174
 [3] offsets
   @ ~/.julia/packages/Stencils/FYZFU/src/stencils/rectangle.jl:35
 [4] multiple call sites
   @ unknown:0
Reason: unsupported call to an unknown function (call to julia.pop_gc_frame)
Stacktrace:
 [1] multiple call sites
   @ unknown:0
Hint: catch this exception as `err` and call `code_typed(err; interactive = true)` to introspect the erronous code with Cthulhu.jl
Stacktrace:
  [1] check_ir(job::GPUCompiler.CompilerJob{GPUCompiler.MetalCompilerTarget, Metal.MetalCompilerParams}, args::LLVM.Module)
    @ GPUCompiler ~/.julia/packages/GPUCompiler/2mJjc/src/validation.jl:147
  [2] macro expansion
    @ ~/.julia/packages/GPUCompiler/2mJjc/src/driver.jl:440 [inlined]
  [3] macro expansion
    @ ~/.julia/packages/TimerOutputs/RsWnF/src/TimerOutput.jl:253 [inlined]
  [4] macro expansion
    @ ~/.julia/packages/GPUCompiler/2mJjc/src/driver.jl:439 [inlined]
  [5] 
    @ GPUCompiler ~/.julia/packages/GPUCompiler/2mJjc/src/utils.jl:92
  [6] emit_llvm
    @ ~/.julia/packages/GPUCompiler/2mJjc/src/utils.jl:86 [inlined]
  [7] 
    @ GPUCompiler ~/.julia/packages/GPUCompiler/2mJjc/src/driver.jl:129
  [8] codegen
    @ ~/.julia/packages/GPUCompiler/2mJjc/src/driver.jl:110 [inlined]
  [9] 
    @ GPUCompiler ~/.julia/packages/GPUCompiler/2mJjc/src/driver.jl:106
 [10] compile
    @ ~/.julia/packages/GPUCompiler/2mJjc/src/driver.jl:98 [inlined]
 [11] #45
    @ ~/.julia/packages/Metal/lnkVP/src/compiler/compilation.jl:57 [inlined]
 [12] JuliaContext(f::Metal.var"#45#46"{GPUCompiler.CompilerJob{GPUCompiler.MetalCompilerTarget, Metal.MetalCompilerParams}})
    @ GPUCompiler ~/.julia/packages/GPUCompiler/2mJjc/src/driver.jl:47
 [13] compile(job::GPUCompiler.CompilerJob)
    @ Metal ~/.julia/packages/Metal/lnkVP/src/compiler/compilation.jl:56
 [14] actual_compilation(cache::Dict{…}, src::Core.MethodInstance, world::UInt64, cfg::GPUCompiler.CompilerConfig{…}, compiler::typeof(Metal.compile), linker::typeof(Metal.link))
    @ GPUCompiler ~/.julia/packages/GPUCompiler/2mJjc/src/execution.jl:125
 [15] cached_compilation(cache::Dict{…}, src::Core.MethodInstance, cfg::GPUCompiler.CompilerConfig{…}, compiler::Function, linker::Function)
    @ GPUCompiler ~/.julia/packages/GPUCompiler/2mJjc/src/execution.jl:103
 [16] macro expansion
    @ ~/.julia/packages/Metal/lnkVP/src/compiler/execution.jl:162 [inlined]
 [17] macro expansion
    @ ./lock.jl:267 [inlined]
 [18] mtlfunction(f::typeof(Stencils.gpu_mapstencil_kerneln0!), tt::Type{Tuple{…}}; name::Nothing, kwargs::@Kwargs{})
    @ Metal ~/.julia/packages/Metal/lnkVP/src/compiler/execution.jl:157
 [19] mtlfunction(f::typeof(Stencils.gpu_mapstencil_kerneln0!), tt::Type{Tuple{…}})
    @ Metal ~/.julia/packages/Metal/lnkVP/src/compiler/execution.jl:155
 [20] macro expansion
    @ ~/.julia/packages/Metal/lnkVP/src/compiler/execution.jl:77 [inlined]
 [21] (::KernelAbstractions.Kernel{…})(::Function, ::Vararg{…}; ndrange::Nothing, workgroupsize::Nothing)
    @ Metal.MetalKernels ~/.julia/packages/Metal/lnkVP/src/MetalKernels.jl:105
 [22] Kernel
    @ ~/.julia/packages/Metal/lnkVP/src/MetalKernels.jl:101 [inlined]
 [23] mapstencil!(::Function, ::MtlMatrix{…}, ::StencilArray{…})
    @ Stencils ~/.julia/packages/Stencils/FYZFU/src/mapstencil.jl:81
 [24] top-level scope
    @ ~/Documents/GitHub/ItsLivePlayground.jl/src/Test_Stencils:99
Some type information was truncated. Use `show(err)` to see complete types.
```

**Comment by rafaqz** [comment]
Ok. I  have to set up CUDA on my laptop again with a new OS 😅 will debug then

(Would be good to have a GPU in CI...)

**Comment by rafaqz** [comment]
This really should work on GPU now, it was probably that `@assert`.

I'll merge for now, make an issue if its still broken.

**Pull request merged**